clippy_lints/src/ambiguous_method_calls.rs

```diff
@@ -0,0 +1,179 @@
+            let parent_item = cx.tcx.hir().get_parent_item(hir_id);
+
+            // Calling type_of on a method's default impl causes an ICE
+            if let Some(hir::Node::Item(item)) = cx.tcx.hir().find(hir::HirId::from(parent_item)) {
+                if let hir::ItemKind::Trait(..) = item.kind {
+                    return;
+                }
+            }
+
+            let parent_type = cx.tcx.type_of(parent_item.to_def_id()).skip_binder();
+            let parent_ty_str = format!("{parent_type}");
+
+            insert_method(is_trait_impl, parent_type, ident);
+
+            if has_ambiguous_name(parent_type, ident) {
+                let trait_methods = trait_methods().lock().unwrap();
+                let struct_methods = struct_methods().lock().unwrap();
+
+                span_lint(
+                    cx,
+                    AMBIGUOUS_METHOD_CALLS,
+                    trait_methods.get(&(parent_ty_str.clone(), ident.name)).unwrap().span(),
+                    "ambiguous trait method name",
+                );
+                span_lint_and_help(
+                    cx,
+                    AMBIGUOUS_METHOD_CALLS,
+                    struct_methods.get(&(parent_ty_str, ident.name)).unwrap().span(),
+                    "ambiguous struct method name",
+                    None,
+                    "consider renaming the struct impl's method",
+                );
+            }
+        }
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>) {
+        if let hir::ExprKind::MethodCall(path, receiver, _, call_span) = &expr.kind {
+            let recv_ty = cx.typeck_results().expr_ty(receiver).peel_refs();
+            let recv_ty_str = format!("{recv_ty}");
+
+            let struct_methods = struct_methods().lock().unwrap();
+            let trait_methods = trait_methods().lock().unwrap();
+            if struct_methods.contains_key(&(recv_ty_str.clone(), path.ident.name))
+                && trait_methods.contains_key(&(recv_ty_str, path.ident.name))
+            {
+                span_lint_and_help(
+                    cx,
+                    AMBIGUOUS_METHOD_CALLS,
+                    *call_span,
+                    "ambiguous struct method call",
+                    None,
+                    "consider renaming the struct impl's method or explicitly qualifying the call site",
+                );
+            }
+        }
+    }
+}
+
+fn has_ambiguous_name(ty: Ty<'_>, ident: Ident) -> bool {
+    let ty_str = format!("{ty}");
+    let trait_methods = trait_methods().lock().unwrap();
+    let struct_methods = struct_methods().lock().unwrap();
+
+    trait_methods.contains_key(&(ty_str.clone(), ident.name)) && struct_methods.contains_key(&(ty_str, ident.name))
+}
+
+fn trait_methods() -> &'static Mutex<FxHashMap<(String, Symbol), SpanData>> {
+    static NAMES: OnceLock<Mutex<FxHashMap<(String, Symbol), SpanData>>> = OnceLock::new();
```

Comment:
Is there a reason for using `static`s instead of storing the map directly in the `AmbiguousMethodCalls` struct? The lint structs can have fields and take `&mut self` in their check methods so its state can be accessed and mutated at any point, which means the `Mutex` could go too.

A problem with the current approach is that it depends on the order in which it visits the items. For example, if a `Base.ambiguous();` expression appears before an ambiguous method implementation, I think it wouldn't lint on it because clippy hasn't seen it yet.

I wonder if maintaining these name maps is necessary though. Rustc already has its own maps for trait/inherent impls per crate for its analysises and typechecking, and they can be accessed through the `TyCtxt`.

For example, there's the [`all_local_trait_impls`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.all_local_trait_impls) query that returns all trait impls for a given type's `DefId`. Maybe when visiting a method in an inherent impl, this could be used to look for traits that have methods with the same name.
I'm not sure if there's a better query to use for this... A problem that I can see here is that it probably wouldn't lint when calling ambiguous unqualified trait methods from foreign crates.

And for linting method call expressions, it could perhaps use `all_local_trait_impls` to get all trait impls + [`inherent_impls`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.inherent_impls) for the normal impls, then check if there's duplicate entries between the two.

But maybe having these two maps is faster over having to go through the impls and compare a bunch of `Symbol`s that it justifies having them.

Reply:
> so linting happens only after visiting all implementations and exprs.

Yep that's a neat way to solve it, FWIW [some other lints follow the same idea](https://github.com/rust-lang/rust-clippy/blob/ddbe1100813d75898d6cde56a498813ac5799e9d/clippy_lints/src/unused_async.rs#L159). For example `unused_async` doesn't lint immediately once it sees an async function without `.await` exprs, but instead stores the relevant stuff for later + delays linting until `check_crate_post` so that we can be sure there aren't any path exprs to an async function in a context that requires a `fn() -> impl Future`.

If going with manual maps, there's one subtle but important thing: the iteration order of a HashMap is not stable, so the order in which it then emits lints can "randomly" change. That makes it annoying for e.g. ui tests where it checks for the exact stderr literally and a different order could then make the tests fail.
I ran into the exact same problem before ^^ see the relevant comment by Jarcho here https://github.com/rust-lang/rust-clippy/pull/11200#discussion_r1271338932

The `inherent_methods` field could just be a `Vec`, right? Seems like it's only inserted into and iterated over.


> As for using rustc's own maps, I'm currently trying to figure something out with the all_traits and all_local_trait_impls queries, and iterating through methods in trait impls to find those that are ambiguous with inherent methods (i.e., same name and same parent type) with something like this in check_fn

Oh huh I was under the impression that the key of `all_local_trait_impls` was the DefId of the type and not the trait, which would have been way more convenient (wouldn't require going through all of the traits). Definitely changes the way I think about this now. All of the trait impl queries on the tyctxt seem to search by a trait id (unless I missed something), which might make this more complicated then..

Reply:
> The inherent_methods field could just be a `Vec`, right? Seems like it's only inserted into and iterated over.

Yeah, I think I was initially worried about getting the correct spans for each inherent impl (hence the `self.inherent_methods.get(k)`), but now that I think about it, this could also be accomplished using something like a `Vec<(String, Symbol, Span)>` instead of a `FxHashMap`--I believe this also solves the random ordering problem.

> I was under the impression that the key of `all_local_trait_impls`'s map was the DefId of the type and not the trait

That tripped me up as well when I was working on the rustc-based approach. I think this fact would also make linting call sites a lot more difficult, though I haven’t gotten around to writing that yet—do you think it’s still worth pursuing, or should we just stick with manual maps? I’m personally more in favor of manual maps—I feel like querying `all_local_trait_impls`, `all_traits`, and `inherent_impls` all at the same time when linting method calls could get complicated pretty quickly.